contrib/go/src/python/pants/contrib/go/subsystems/protoc_gen_go.py

```diff
@@ -0,0 +1,69 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+import logging
+import os
+
+from pants.backend.codegen.protobuf.subsystems.protoc import Protoc
+from pants.base.workunit import WorkUnitLabel
+from pants.scm.git import Git
+from pants.subsystem.subsystem import Subsystem, SubsystemError
+from pants.util.dirutil import safe_mkdir
+from pants.util.memo import memoized_method
+
+from pants.contrib.go.subsystems.go_distribution import GoDistribution
+
+
+logger = logging.getLogger(__name__)
+
+
+class ProtocGenGo(Subsystem):
+  """A compiled protobuf plugin that generates Go code.
+
+  For details, see https://github.com/golang/protobuf
+  """
+  options_scope = 'protoc-gen-go'
+
+  @classmethod
+  def register_options(cls, register):
+    super(ProtocGenGo, cls).register_options(register)
+    register('--version', default='v1.1.0',
+             help='Version of protoc-gen-go plugin to use when generating code')
+
+  @classmethod
+  def subsystem_dependencies(cls):
+    return super(ProtocGenGo, cls).subsystem_dependencies() + (Protoc.scoped(cls), GoDistribution,)
+
+  @memoized_method
+  def select(self, context):
+    self.get_options()
+    workdir = os.path.join(self.get_options().pants_workdir, self.options_scope,
+                           'versions', self.get_options().version)
+    tool_path = os.path.join(workdir, 'bin/protoc-gen-go')
+
+    if not os.path.exists(tool_path):
```

Comment:
There is a potential failure mode here where if the checkout or install fail, we're left with a corrupted checkout.

To avoid that, you can use `safe_concurrent_creation` to create in a temp directory, which can be renamed into the final location only if the checkout completes successfully: https://github.com/pantsbuild/pants/blob/4757f9677adfec90a399cb92227fe4686faefcc2/src/python/pants/util/dirutil.py#L310-L319

The entire body of this if statement would be inside the contextmanager.

Reply:
I _think_ this will recover from a failed checkout, etc. because if the binary doesn't exist, we delete the workdir and try again:

```
safe_mkdir(workdir, clean=True)
```

That said, this context manager is handy! I'll switch things over.